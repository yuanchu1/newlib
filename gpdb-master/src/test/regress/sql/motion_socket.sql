-- The following test checks if the correct number and type of sockets are
-- created for motion connections both on QD and QE backends for the same
-- gp_session_id. Additionally we check if the source address used for creating
-- the motion sockets is equal to gp_segment_configuration.address.


-- start_matchsubs
-- m/^INFO:  Checking postgres backend postgres:.*/
-- s/^INFO:  Checking postgres backend postgres:.*/INFO:  Checking postgres backend postgres: XXX/
-- end_matchsubs
CREATE FUNCTION check_motion_sockets()
    RETURNS VOID as $$
import psutil, socket

# Create a temporary table to create a writer gang
plpy.execute("CREATE TEMP TABLE motion_socket_force_create_gang(i int);")

# Since we have slightly different logic in constructing the source address of
# motion sockets for singleton readers on the QD, create a singleton reader on
# the QD as well.
plpy.execute("SELECT * FROM motion_socket_force_create_gang, pg_database;")

# We expect different number of sockets to be created for different
# interconnect types
# UDP: See calls to setupUDPListeningSocket in InitMotionUDPIFC
# TCP/PROXY: See call to setupTCPListeningSocket in InitMotionTCP
res = plpy.execute("SELECT current_setting('gp_interconnect_type');", 1)
ic_type = res[0]['current_setting']
if ic_type in ['tcp', 'proxy']:
    expected_socket_count_per_segment = 1
    expected_socket_kind = socket.SocketKind.SOCK_STREAM
elif ic_type=='udpifc':
    expected_socket_count_per_segment = 2
    expected_socket_kind = socket.SocketKind.SOCK_DGRAM
else:
    plpy.error('Unrecognized gp_interconnect_type {}.'.format(ic_type))

# Since this test is run on a single physical host we assume that all segments
# have the same gp_segment_configuration.address
res = plpy.execute("SELECT address FROM gp_segment_configuration;", 1)
hostip = socket.gethostbyname(res[0]['address'])

res = plpy.execute("SELECT pid from gp_backend_info();")
pids_to_check = [r['pid'] for r in res]

for pid in pids_to_check:
    # We iterate through all backends related to current session
    motion_socket_count = 0
    process = psutil.Process(pid)
    plpy.info('Checking postgres backend {}'.format(process.cmdline()[0]))
    for conn in process.connections():
        if conn.type == expected_socket_kind and conn.raddr == () \
        and conn.laddr.ip == hostip:
            motion_socket_count += 1

    if motion_socket_count != expected_socket_count_per_segment:
        plpy.error('Expected {} motion sockets but found {}. '\
        'For backend process {}. connections= {}'\
        .format(expected_socket_count_per_segment, process,\
        motion_socket_count, process.connections()))


$$ LANGUAGE plpython3u;
SELECT check_motion_sockets();
