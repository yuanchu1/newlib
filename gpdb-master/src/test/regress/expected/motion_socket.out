-- The following test checks if the correct number and type of sockets are
-- created for motion connections both on QD and QE backends for the same
-- gp_session_id. Additionally we check if the source address used for creating
-- the motion sockets is equal to gp_segment_configuration.address.
-- start_matchsubs
-- m/^INFO:  Checking postgres backend postgres:.*/
-- s/^INFO:  Checking postgres backend postgres:.*/INFO:  Checking postgres backend postgres: XXX/
-- end_matchsubs
CREATE FUNCTION check_motion_sockets()
    RETURNS VOID as $$
import psutil, socket

# Create a temporary table to create a writer gang
plpy.execute("CREATE TEMP TABLE motion_socket_force_create_gang(i int);")

# Since we have slightly different logic in constructing the source address of
# motion sockets for singleton readers on the QD, create a singleton reader on
# the QD as well.
plpy.execute("SELECT * FROM motion_socket_force_create_gang, pg_database;")

# We expect different number of sockets to be created for different
# interconnect types
# UDP: See calls to setupUDPListeningSocket in InitMotionUDPIFC
# TCP/PROXY: See call to setupTCPListeningSocket in InitMotionTCP
res = plpy.execute("SELECT current_setting('gp_interconnect_type');", 1)
ic_type = res[0]['current_setting']
if ic_type in ['tcp', 'proxy']:
    expected_socket_count_per_segment = 1
    expected_socket_kind = socket.SocketKind.SOCK_STREAM
elif ic_type=='udpifc':
    expected_socket_count_per_segment = 2
    expected_socket_kind = socket.SocketKind.SOCK_DGRAM
else:
    plpy.error('Unrecognized gp_interconnect_type {}.'.format(ic_type))

# Since this test is run on a single physical host we assume that all segments
# have the same gp_segment_configuration.address
res = plpy.execute("SELECT address FROM gp_segment_configuration;", 1)
hostip = socket.gethostbyname(res[0]['address'])

res = plpy.execute("SELECT pid from gp_backend_info();")
pids_to_check = [r['pid'] for r in res]

for pid in pids_to_check:
    # We iterate through all backends related to current session
    motion_socket_count = 0
    process = psutil.Process(pid)
    plpy.info('Checking postgres backend {}'.format(process.cmdline()[0]))
    for conn in process.connections():
        if conn.type == expected_socket_kind and conn.raddr == () \
        and conn.laddr.ip == hostip:
            motion_socket_count += 1

    if motion_socket_count != expected_socket_count_per_segment:
        plpy.error('Expected {} motion sockets but found {}. '\
        'For backend process {}. connections= {}'\
        .format(expected_socket_count_per_segment, process,\
        motion_socket_count, process.connections()))


$$ LANGUAGE plpython3u;
SELECT check_motion_sockets();
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
INFO:  Checking postgres backend postgres:  7000, vanjared regression 127.0.0.1(59545) con1812 cmd2 SELECT
INFO:  Checking postgres backend postgres:  7002, vanjared regression 127.0.0.1(59550) con1812 seg0 idle in transaction
INFO:  Checking postgres backend postgres:  7003, vanjared regression 127.0.0.1(59551) con1812 seg1 idle in transaction
INFO:  Checking postgres backend postgres:  7004, vanjared regression 127.0.0.1(59552) con1812 seg2 idle in transaction
INFO:  Checking postgres backend postgres:  7000, vanjared regression 127.0.0.1(59553) con1812 seg-1 idle
 check_motion_sockets 
----------------------
 
(1 row)

